[AIRFLOW-4495] allow externally triggered dags to run for future 'Exe…

tooptoop4 opened the pull request:
allow externally triggered dags to run for future 'Execution date'

Make sure you have checked _all_ steps below.

### Jira AIRFLOW-4495

- [x] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-XXX
  - In case you are fixing a typo in the documentation you can prepend your commit with \[AIRFLOW-XXX\], code changes always need a Jira issue.
  - In case you are proposing a fundamental code change, you need to create an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)).
  - In case you are adding a dependency, check if the license complies with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:

Allow future dated execution dates

### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:
Tested in production env for 2 weeks

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain docstrings that explain what it does
  - If you implement backwards incompatible changes, please leave a note in the [Updating.md](https://github.com/apache/airflow/blob/master/UPDATING.md) so we can assign it to a appropriate release

### Code Quality

- [x] Passes `flake8`

Commit:
[AIRFLOW-4495] allow externally triggered dags to run for future 'Execution date'

Commit:
[AIRFLOW-4428] Externally triggered DagRun marked as 'success' state but all tasks for it are 'None' state

Commit:
Revert "[AIRFLOW-4428] Externally triggered DagRun marked as 'success' state but all tasks for it are 'None' state"

This reverts commit e4aed3f6e560a16ceb381ebcb3fe10be810af92d.

XD-DENG commented on airflow/jobs/scheduler_job.py:
```diff
@@ -684,13 +684,6 @@ def _process_task_instances(self, dag, task_instances_list, session=None):
-                    "Execution date is in future: %s",
-                    run.execution_date
-                )
-                continue
```
This part is the scheduler code. If your DagRun's execution_date is in the future, for sure it should not be considered for execution. This existing chunk is behaving correctly to me, so I don't quite get the idea why you want to remove this check.

- If you want to run a job immediately, you should give execution_date correctly
- If you're intending to immediately execute a DagRun whose execution_date is in the future, I suspect there is some fundamental misunderstanding.

tooptoop4 replied:
my dag will do ETL on a file where the filename contains the executiondate. I receive both present year data and 'forecast' data where filename is next year (so i do 2 externally triggered dagruns, 1 per date of filename). i want to run the dag for the forecast executiondate immediately

yuqian90 replied:
@XD-DENG I actually disagree with the statement "If your DagRun's execution_date is in the future, for sure it should not be considered for execution"

We should think about what execution_date is. It is definitely NOT the datetime tasks in the DAG starts running. So there's no reason to make the scheduler ignore tasks with an execution_date in the future. In the current implementation, if execution_date is T, and we set Dag.schedule to a cron expression, the first task in the DAG can run at (T + one schedule_interval). See https://airflow.apache.org/concepts.html
"The time Airflow triggers a DAG should equal execution_date plus one schedule interval if cron expression is used: The scheduler runs your job one schedule_interval AFTER the start date, at the END of the period"

If we set DAG.schedule to None and use external trigger to trigger the DAG, naturally we should expect the scheduler to start considering the tasks for execution once the DAG is triggered, even the execution_date may be a future date. In the current implementation, this is not the case. If execution_date is 20190801, the earliest time the tasks in the DAG can be considered by the scheduler is 20190801 00:00 UTC because of this line of code here. So if someone uses external trigger to trigger the dag at 20190731 23:00 UTC with execution_date 20190801, the tasks will not be considered for execution until 20190801 00:00 UTC. 

This is very problematic if you are working with multiple timezones.  E.g if you are in Asia/Tokyo and you want the jinja templates to evaluate ds_nodash to 20190801, you need to make execution_date 20190801. But if you do that, the DAG can never start running before 20190801 00:00 UTC which is 20190801 09:00 Tokyo time. So if you want something to run at 20190801 8am Tokyo time with execution_date 20190801, it is not possible (even if you trigger the DAG before 9am externally). 

Removing this check for DAG.schedule == None seems to be a move in the right direction. Although this pull request needs a lot more testing to make sure it does the right thing. And it needs to give people some warning before this change goes out.

Commit:
remove unused import

Commit:
[AIRFLOW-4428] fix test

tooptoop4 committed:
Config flag to skip SLA checks

Some airflow users have no use of managing SLAs within airflow. I believe the scheduling process should be as fast as possible and not do unnecessary logging, the current IF statement is slower than a boolean flag and produces a redundant log.   https://jira.apache.org/jira/browse/AIRFLOW-5149

tooptoop4 committed:
Update default_airflow.cfg
